src/logger.py

```diff
@@ -0,0 +1,57 @@
+import logging
+from typing import Iterable, Union, List, Any, Callable, Tuple
+from rich.logging import RichHandler
+from .config import CONFIG_DEFAULTS as config
+
+FORMAT = "%(message)s"
+
+logging.basicConfig(
+                level=logging.NOTSET, format="%(message)s", datefmt="[%X]", handlers=[RichHandler(rich_tracebacks=config.outputs.rich_tracebacks)]
+                )
+
+class Logger:
+    def __init__(self, name, level: Union[int, str] = logging.NOTSET,  message_format = "%(message)s", date_format = "[%X]"):
+        self.log = logging.getLogger(name)
+        self.log.setLevel(level)
+        self.log.__format__ = message_format
+        self.log.__date_format__ = date_format
+        
+    def stringify(func):
+        def inner(self, method_type: str, *msg: object, sep=' '):
+            nmsg = []
+            for v in msg:
+                if not isinstance(v, str):
+                    v = str(v)
+                nmsg.append(v)
+            return func(self, method_type, *nmsg, sep=sep)
+        return inner
+    
+    # set stack level to 3 so that the caller of this function is logged, not this function itself.
+    # stack-frame - self.log.debug - logutil - stringify - log method - caller
+    @stringify 
+    def logutil(self, method_type: str, *msg: object, sep=' ') -> None:
+        func = getattr(self.log, method_type, None)
+        if not func:
+            raise AttributeError(f"Logger has no method {method_type}")
+        return func(sep.join(msg), stacklevel=4)
```

Comment:
add spaces around '=' at all places e.g. `stacklevel = 4`

Reply:
Spaces are not recommended around keyword arguments according to PEP.

Reply:
odd, but true. cool